lib/dynamodb/dynamodb_request_queue.c

```diff
@@ -243,6 +221,63 @@ extractcapacity(struct http_response * res, double * pcap)
 	return (-1);
 }
 
+/* Extract a DynamoDB error string from an HTTP 400 response. */
+static int
+extracterror(struct http_response * res, char ** errstr)
+{
+	const uint8_t * buf = res->body;
+	const uint8_t * end = res->body + res->bodylen;
+	size_t len;
+
+	/* Look for __type. */
+	buf = json_find(buf, end, "__type");
+
+	/*
+	 * We should be pointing at the opening double-quote of a string;
+	 * move past it.
+	 */
+	if (buf[0] != '"')
+		goto bad;
+	buf++;
+
+	/* Find a '#' inside the string and move past it. */
+	while (buf[0] != '#') {
+		if (buf[0] == '"')
+			goto bad;
+		buf++;
+	}
+	buf++;
+
+	/* The error string is everything from here to the closing '"'. */
+	for (len = 0; &buf[len] < end; len++) {
+		if (buf[len] == '"')
+			break;
+	}
+	if (&buf[len] == end)
+		goto bad;
+
+	/* Duplicate the error string. */
```

Comment:
This works, but why not use `strndup()`?

Reply:
Because we already know the length here.  Ok, this isn't a hugely performance-critical path, but `strndup` is slower than `malloc+memcpy` when we already know how much to copy.